snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java

```diff
@@ -15,169 +15,847 @@
-		final Multimap<String, QueryExpression> singleConceptInclusions = FluentIterable.from(inclusions)
-				.filter(ex -> isSingleConceptExpression(eclCache, ex.getQuery()))
-				.index(ex -> toSingleConceptId(eclCache, ex.getQuery()));
+		final QueryExpressionDiff diff = QueryExpressionDiff.create(
+			context, 
+			inclusions, 
+			exclusions, 
+			optimizedInclusions, 
+			optimizedExclusions);
+
+		log.info("Optimize resulted in {} addition(s) to inclusion clauses, {} addition(s) to exclusion clauses and {} removal(s)", 
+			diff.getAddToInclusion().size(),
+			diff.getAddToExclusion().size(),
+			diff.getRemove().size());
 		
-		// if there are no single concept inclusions to optimize, exit early
-		if (singleConceptInclusions.isEmpty()) {
+		if (diff.getAddToExclusion().isEmpty() && diff.getAddToExclusion().isEmpty() && diff.getRemove().isEmpty()) {
 			return QueryExpressionDiffs.EMPTY;
 		}
 		
-		// Record the ancestors (both direct and indirect) of each single concept inclusion
-		final Multimap<String, QueryExpression> membersByAncestor = HashMultimap.create();
-		
-		final int partitionSize = context.service(RepositoryConfiguration.class)
-			.getIndexConfiguration()
-			.getTermPartitionSize();
-		
-		Iterables.partition(singleConceptInclusions.keySet(), partitionSize).forEach(batchIds -> {
-			SnomedRequests.prepareSearchConcept()
-				.filterByIds(batchIds)
-				.setLimit(batchIds.size())
-				.build()
-				.execute(context)
-				.forEach(child -> {
-					final Collection<QueryExpression> childExpressions = singleConceptInclusions.get(child.getId());
-					final List<String> parentIds = child.getParentIdsAsString();
-					final List<String> ancestorIds = child.getAncestorIdsAsString();
-					
-					parentIds.forEach(parentId -> {
-						if (!IComponent.ROOT_ID.equals(parentId) && !Concepts.ROOT_CONCEPT.equals(parentId)) {
-							membersByAncestor.putAll(parentId, childExpressions);
-						}
-					});
-					
-					ancestorIds.forEach(ancestorId -> {
-						if (!IComponent.ROOT_ID.equals(ancestorId) && !Concepts.ROOT_CONCEPT.equals(ancestorId)) {
-							membersByAncestor.putAll(ancestorId, childExpressions);
-						}
-					});
-				});
-		});
+		return addLabelsToDiff(context, locales, diff);
+	}
+
+	private Set<String> optimizeInclusions(final BranchContext context, final Collection<QueryExpression> inclusions) {
+		optimizedInclusions = newArrayList();
+
+		// Pinned inclusions are checked for redundancy but otherwise should appear unmodified in optimized output 
+		final List<QueryExpression> pinnedInclusions = optimizePinned(context, inclusions);
+		log.info("Found {} unique pinned inclusion(s)", pinnedInclusions.size());
+		applyInclusions(context, pinnedInclusions);
+
+		// See if concepts still to be processed can be converted to "* : c1 = c2" expressions
+		final SnomedRelationshipStats inclusionRelationshipStats = SnomedRelationshipStats.create(
+			context, 
+			conceptsToInclude, 
```

Comment:
This line could be `conceptSet` here to be consistent with the other calls and I think by doing that the comment on line 365 is not needed anymore.

Reply:
IIRC this distinction was made in the reference algorithm as well for some reason. Ancestors are allowed to cover an included concept multiple times.